src/Compilers/CSharp/Portable/Emitter/Model/PEModuleBuilder.cs

```diff
@@ -208,6 +208,108 @@ internal sealed override IEnumerable<Cci.INestedTypeDefinition> GetSynthesizedNe
             return null;
         }
 
+        public sealed override MultiDictionary<Cci.ITypeDefinition, Cci.DebugSourceDocument> GetTypeToDebugDocumentMap(EmitContext context)
+        {
+            var result = new MultiDictionary<Cci.ITypeDefinition, Cci.DebugSourceDocument>(Cci.SymbolEquivalentEqualityComparer.Instance);
+
+            var namespacesAndTypesToProcess = new Stack<NamespaceOrTypeSymbol>();
+            namespacesAndTypesToProcess.Push(SourceModule.GlobalNamespace);
+            while (namespacesAndTypesToProcess.Count > 0)
+            {
+                var symbol = namespacesAndTypesToProcess.Pop();
+
+                switch (symbol.Kind)
+                {
+                    case SymbolKind.Namespace:
+                        var location = GetSmallestSourceLocationOrNull(symbol);
+
+                        // filtering out synthesized symbols not having real source 
+                        // locations such as anonymous types, etc...
+                        if (location != null)
+                        {
+                            foreach (var member in symbol.GetMembers())
+                            {
+                                switch (member.Kind)
+                                {
+                                    case SymbolKind.Namespace:
+                                    case SymbolKind.NamedType:
+                                        namespacesAndTypesToProcess.Push((NamespaceOrTypeSymbol)member);
+                                        break;
+                                    default:
+                                        throw ExceptionUtilities.UnexpectedValue(member.Kind);
+                                }
+                            }
+                        }
+                        break;
+                    case SymbolKind.NamedType:
+                        var methodDocumentList = GetDocumentsForMethods(symbol, context, out var allMethodsHaveIL);
+
+                        // If all of the methods have IL then the document data we already have is enough, but if not
+                        // then we want to add document info for the type.
+                        if (!allMethodsHaveIL)
```

Comment:
I think this can be done better. What we need is to check if the set of documents that contain a part of the type's declaration is a strict superset of the set of documents that contain any method of the type that has a method body. If it is so then we need to emit the list of documents the type spans. Otherwise we don't need to.

Reply:
Do you mean we should just get a list of documents for the type, and a list of documents for any method of the type, and then emit the items in the first list that aren't in the second? Because that's essentially what this code is doing, it just has an extra optimisation where it can avoid bothering to create the first list if we know that all methods have IL.

I'm happy to remove that optimisation if you think its not warranted, as it would certainly improve readability.

Reply:
I don't think this is a correct optimization. Consider:

```C#
partial class C { C() {} }
```

```C#
partial class C { int F; }
```

in this case all methods of C have body with IL, but they don't cover all documents C might be declared in.

Reply:
Thank you, that makes sense.

Interestingly your scenario actually works now, because the synthesized constructor has no body, but changing it to add a field initializer and moving the field to the first document, fails with the current code.

Reply:
There should be no synthesized ctor in my example. in any case, you see the issue

Reply:
Tell that to the debugger 😛